multiversx/mx-chain-go
# Feat/balance data tries

## Reasoning behind the pull request
- The keys at which values are saved in a data trie are not random. Because of this, the data tries are not balanced, thus resulting in more intermediary nodes. 
  
## Proposed changes
- When saving some data in the data trie, do not save at key but rather at hash(key). This way, the keys will be random, resulting in balanced data tries.
- In order for this change to be backwards compatible, add `Version` field to trie nodes. Nodes that are accessed will be automatically migrated to the new version (where the storage key is hash(key))
- Add a built in function that will migrate data trie nodes when it is called
- Add a new API endpoint which will return true if the data trie of a specified account is fully migrated.

## Testing procedure
- Deploy a contract with some data before the activation of this feature. After the activation, access all the keys and check that they are automatically migrated.
- Deploy a contract with some data before the activation of this feature. Use the built in function to migrate the data trie after the feature is active.
- Check that the data that is added after the feature is active is saved in the trie by using the new version.

## Pre-requisites

Based on the [Contributing Guidelines](CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?
